#**Inteligência Artificial e Previsões**

* Fiz uma análise de Score de cliente em um banco. Imagine que você trabalha em um banco e tem uma base de dados de 100.000 clientes e precisa verificar o score (pontuação) do cliente para saber se é um bom cliente ou
não.

* Isso vai definir se o banco vai emprestar dinheiro, se ele vai ter crédito entre outros benefícios dentro do banco. Então o nosso objetivo é fazer um tratamento na base de dados e criar alguns algoritmos de classificação e verificar qual deles é o melhor.

* Com isso nós vamos conseguir fazer uma previsão dos clientes para saber se ele vai ser um cliente com um bom score ou não.

* Isso é muito importante, pois o banco não vai querer um cliente que não paga e que fica devendo não é mesmo?

**Os passo que realizei são**:

• Importar a base de dados

• Verificar informações vazias

• Fazer o tratamento na base de dados

• Selecionar as colunas de treino para o modelo

• Treinar 2 modelos

• Verificar o melhor modelo

• Verificar quais as características mais importantes para definir o score do cliente

#**1. Importar a base de dados**

In [12]:
import pandas as pd

tabela = pd.read_csv("clientes.csv") # importa a base de dados
display(tabela)

id_cliente  mes  idade  profissao  salario_anual  num_contas  \
0            3392    1   23.0  cientista       19114.12         3.0   
1            3392    2   23.0  cientista       19114.12         3.0   
2            3392    3   23.0  cientista       19114.12         3.0   
3            3392    4   23.0  cientista       19114.12         3.0   
4            3392    5   23.0  cientista       19114.12         3.0   
...           ...  ...    ...        ...            ...         ...   
99995       37932    4   25.0   mecanico       39628.99         4.0   
99996       37932    5   25.0   mecanico       39628.99         4.0   
99997       37932    6   25.0   mecanico       39628.99         4.0   
99998       37932    7   25.0   mecanico       39628.99         4.0   
99999       37932    8   25.0   mecanico       39628.99         4.0   

       num_cartoes  juros_emprestimo  num_emprestimos  dias_atraso  ...  \
0              4.0               3.0              4.0          3.0  ...   
1              4.0               3.0              4.0          3.0  ...   
2              4.0               3.0              4.0          3.0  ...   
3              4.0               3.0              4.0          5.0  ...   
4              4.0               3.0              4.0          6.0  ...   
...            ...               ...              ...          ...  ...   
99995          6.0               7.0              2.0         23.0  ...   
99996          6.0               7.0              2.0         18.0  ...   
99997          6.0               7.0              2.0         27.0  ...   
99998          6.0               7.0              2.0         20.0  ...   
99999          6.0               7.0              2.0         18.0  ...   

       idade_historico_credito  investimento_mensal  \
0                        265.0            21.465380   
1                        266.0            21.465380   
2                        267.0            21.465380   
3                        268.0            21.465380   
4                        269.0            21.465380   
...                        ...                  ...   
99995                    378.0            24.028477   
99996                    379.0            24.028477   
99997                    380.0            24.028477   
99998                    381.0            24.028477   
99999                    382.0            24.028477   

           comportamento_pagamento  saldo_final_mes  score_credito  \
0      alto_gasto_pagamento_baixos       312.494089           Good   
1       baixo_gasto_pagamento_alto       284.629162           Good   
2      baixo_gasto_pagamento_medio       331.209863           Good   
3      baixo_gasto_pagamento_baixo       223.451310           Good   
4       alto_gasto_pagamento_medio       341.489231           Good   
...                            ...              ...            ...   
99995    alto_gasto_pagamento_alto       479.866228           Poor   
99996   alto_gasto_pagamento_medio       496.651610           Poor   
99997    alto_gasto_pagamento_alto       516.809083           Poor   
99998   baixo_gasto_pagamento_alto       319.164979       Standard   
99999   alto_gasto_pagamento_medio       393.673696           Poor   

       emprestimo_carro  emprestimo_casa emprestimo_pessoal  \
0                     1                1                  1   
1                     1                1                  1   
2                     1                1                  1   
3                     1                1                  1   
4                     1                1                  1   
...                 ...              ...                ...   
99995                 1                0                  0   
99996                 1                0                  0   
99997                 1                0                  0   
99998                 1                0                  0   
99999                 1                0                  

**2. Verificar informações vazias**

In [13]:
# verificar se temos valores vazios ou valores reconhecidos em formato errado
print(tabela.info())
print(tabela.columns)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 25 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   id_cliente                100000 non-null  int64  
 1   mes                       100000 non-null  int64  
 2   idade                     100000 non-null  float64
 3   profissao                 100000 non-null  object 
 4   salario_anual             100000 non-null  float64
 5   num_contas                100000 non-null  float64
 6   num_cartoes               100000 non-null  float64
 7   juros_emprestimo          100000 non-null  float64
 8   num_emprestimos           100000 non-null  float64
 9   dias_atraso               100000 non-null  float64
 10  num_pagamentos_atrasados  100000 non-null  float64
 11  num_verificacoes_credito  100000 non-null  float64
 12  mix_credito               100000 non-null  object 
 13  divida_total              100000 non-null  fl

 **3. Fazer o tratamento na base de dados**

In [14]:
from sklearn.preprocessing import LabelEncoder
# vai transformar as colunas de texto em números, ex: profissoes vai sair de cientista, professor, mecanico, etc para 0, 1, 2, etc
codificador = LabelEncoder()

# só não aplicamos na coluna de score_credito que é o nosso objetivo
for coluna in tabela.columns:
    if tabela[coluna].dtype == "object" and coluna != "score_credito":
        tabela[coluna] = codificador.fit_transform(tabela[coluna])

# verificando se realmente todas as colunas foram modificadas
print(tabela.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 25 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   id_cliente                100000 non-null  int64  
 1   mes                       100000 non-null  int64  
 2   idade                     100000 non-null  float64
 3   profissao                 100000 non-null  int64  
 4   salario_anual             100000 non-null  float64
 5   num_contas                100000 non-null  float64
 6   num_cartoes               100000 non-null  float64
 7   juros_emprestimo          100000 non-null  float64
 8   num_emprestimos           100000 non-null  float64
 9   dias_atraso               100000 non-null  float64
 10  num_pagamentos_atrasados  100000 non-null  float64
 11  num_verificacoes_credito  100000 non-null  float64
 12  mix_credito               100000 non-null  int64  
 13  divida_total              100000 non-null  fl

**4. Selecionar as colunas de treino para o modelo**

In [16]:
# escolhendo quais colunas vamos usar para treinar o modelo
# y é a coluna que queremos que o modelo calcule
# x vai todas as colunas que vamos usar para prever o score de credito, não vamos usar a coluna id_cliente porque ela é um numero qualquer que nao ajuda a previsao
x = tabela.drop(["score_credito", "id_cliente"], axis=1)
y = tabela["score_credito"]

from sklearn.model_selection import train_test_split

# separamos os dados em treino e teste. Treino vamos dar para os modelos aprenderem e teste vamos usar para ver se o modelo aprendeu corretamente
x_treino, x_teste, y_treino, y_teste = train_test_split(x, y, test_size=0.3, random_state=1)

**5. Treinar 2 modelos**


In [17]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier

modelo_arvore = RandomForestClassifier() # modelo arvore de decisao
modelo_knn = KNeighborsClassifier() # modelo do KNN (nearest neighbors - vizinhos mais proximos)

# treinando os modelos
modelo_arvore.fit(x_treino, y_treino)
modelo_knn.fit(x_treino, y_treino)

KNeighborsClassifier()

In [18]:
# se o nosso modelo chutasse tudo "Standard", qual seria a acurácia do modelo?
contagem_scores = tabela["score_credito"].value_counts()
print(contagem_scores['Standard'] / sum(contagem_scores))

0.53174



**6. Verificar o melhor modelo**


In [19]:
from sklearn.metrics import accuracy_score

# calculamos as previsoes
previsao_arvore = modelo_arvore.predict(x_teste)
previsao_knn = modelo_knn.predict(x_teste.to_numpy())

# comparamos as previsoes com o y_teste
# esse score queremos o maior (maior acuracia, mas tb tem que ser maior do que o chute de tudo Standard)
print(accuracy_score(y_teste, previsao_arvore))
print(accuracy_score(y_teste, previsao_knn))

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(


0.8242
0.7324


In [24]:
print(modelo_arvore)

RandomForestClassifier()


**7. O número de folhas da árvore treinada**

In [27]:
# Obtenha o número de folhas na árvore treinada
num_folhas = modelo_arvore.estimators_[0].tree_.node_count
print(f"A árvore de decisão treinada possui {num_folhas} folhas.")


A árvore de decisão treinada possui 25039 folhas.


In [28]:
# fazendo novas previsões
novos_clientes = pd.read_csv("novos_clientes.csv")
display(novos_clientes)
for coluna in novos_clientes.columns:
    if novos_clientes[coluna].dtype == "object" and coluna != "score_credito":
        novos_clientes[coluna] = codificador.fit_transform(novos_clientes[coluna])

previsoes = modelo_arvore.predict(novos_clientes)
print(previsoes)

mes  idade   profissao  salario_anual  num_contas  num_cartoes  \
0    1   31.0  empresario      19300.340         6.0          7.0   
1    4   32.0    advogado      12600.445         5.0          5.0   
2    2   48.0  empresario      20787.690         8.0          6.0   

   juros_emprestimo  num_emprestimos  dias_atraso  num_pagamentos_atrasados  \
0              17.0              5.0         52.0                      19.0   
1              10.0              3.0         25.0                      18.0   
2              14.0              7.0         24.0                      14.0   

   ...  taxa_uso_credito idade_historico_credito  investimento_mensal  \
0  ...         29.934186                   218.0             44.50951   
1  ...         28.819407                    12.0              0.00000   
2  ...         34.235853                   215.0              0.00000   

       comportamento_pagamento  saldo_final_mes  emprestimo_carro  \
0  baixo_gasto_pagamento_baixo       312.487689                 1   
1  baixo_gasto_pagamento_medio       300.994163                 0   
2   baixo_gasto_pagamento_alto       345.081577                 0   

  emprestimo_casa  emprestimo_pessoal  emprestimo_credito  \
0               1                   0                   0   
1               0                   0                   0   
2               1                   0                   1   

   emprestimo_estudantil  
0                      0  
1                      1  
2                      0  

[3 rows x 23 columns]

['Poor' 'Good' 'Standard']


**8. Verificar quais as características mais importantes para definir o score do cliente**

In [29]:
# quais as caracteristicas mais importantes para definir o score de credito?
colunas = list(x_teste.columns)
importancia = pd.DataFrame(index=colunas, data=modelo_arvore.feature_importances_)
importancia = importancia * 100
print(importancia)

                                  0
mes                        4.002076
idade                      4.217206
profissao                  3.248371
salario_anual              5.095269
num_contas                 3.466535
num_cartoes                4.874858
juros_emprestimo           8.141210
num_emprestimos            3.139041
dias_atraso                6.457727
num_pagamentos_atrasados   4.597938
num_verificacoes_credito   4.576210
mix_credito                7.961834
divida_total              11.844664
taxa_uso_credito           5.091307
idade_historico_credito    7.107586
investimento_mensal        4.878393
comportamento_pagamento    2.361684
saldo_final_mes            5.438054
emprestimo_carro           0.698383
emprestimo_casa            0.722466
emprestimo_pessoal         0.695747
emprestimo_credito         0.695592
emprestimo_estudantil      0.687847


#**Insights**

- Com isso, foi possível treinar dois modelos de classificação para auxiliar na previsão dos dados. Isso ajudaria a empresa a verificar

- com uma acurácia de 82% quais os clientes que possuem um bom score.
Outro ponto, é que foi possível analisar quais as características mais importantes para definir esse score do cliente.

- Nesse projeto a ideia era diminuir a quantidade de clientes com um score baixo, então consegui criar um modelo para fazer uma previsão e ainda conseguimos analisar quais os pontos importantes precisamos
levar em consideração para análise do score.